codis/pkg/proxy/session.go

```diff
@@ -662,32 +669,56 @@ func (s *Session) handleRequestSlotsMapping(r *Request, d *Router) error {
 	}
 }
 
-func (s *Session) incrOpTotal() {
-	s.stats.total.Incr()
-}
-
-func (s *Session) getOpStats(opstr string) *opStats {
+func (s *Session) getOpStats(opstr string, create bool) *opStats {
+	cmdstats.RLock()
 	e := s.stats.opmap[opstr]
+	cmdstats.RUnlock()
+	if e != nil || !create {
+		return e
+	}
+	cmdstats.Lock()
+	defer cmdstats.Unlock()
+	e = cmdstats.opmap[opstr]
 	if e == nil {
 		e = &opStats{opstr: opstr}
+		for i := 0; i < IntervalNum; i++ {
+			e.delayInfo[i] = &delayInfo{interval: IntervalMark[i]}
+		}
 		s.stats.opmap[opstr] = e
 	}
 	return e
 }
 
 func (s *Session) incrOpStats(r *Request, t redis.RespType) {
-	e := s.getOpStats(r.OpStr)
+
+	if r == nil {
+		return
+	}
+	responseTime := time.Now().UnixNano() - r.ReceiveTime
```

Comment:
<img width="1204" alt="image" src="https://github.com/OpenAtomFoundation/pika/assets/38887641/5b28858d-6b91-4d87-9df8-6e60a4d9c621">
golang 里面时间统计，用 time.Since 好点，使用 时间戳相减，有时候会出现负数，之前我们生产环境遇过这个坑。不知道新版本的 golang  SDK 是否解决了这个问题

Reply:
> <img alt="image" width="1204" src="https://private-user-images.githubusercontent.com/38887641/335924670-5b28858d-6b91-4d87-9df8-6e60a4d9c621.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczODQ5MTcsIm5iZiI6MTcxNzM4NDYxNywicGF0aCI6Ii8zODg4NzY0MS8zMzU5MjQ2NzAtNWIyODg1OGQtNmI5MS00ZDg3LTlkZjgtNmU2MGE0ZDljNjIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjAzVDAzMTY1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5MzBkYzZhNTM1ZjhiODc2ZWVhYmFiZWI5ZTBkMzBmNjA3YWQzMGM3NTc0MmVmYzk0MjQ5ODMyMmJjNzM2YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sod7wWS9iKKcqY_RHAm8PaimUDxCd9noI-Zxux98y3c"> golang 里面时间统计，用 time.Since 好点，使用 时间戳相减，有时候会出现负数，之前我们生产环境遇过这个坑。不知道新版本的 golang 是否解决了这个问题。即，r.ReceiveTime 用 time.Time 来表达，不要使用时间戳
codis没有遇到过

Reply:
> > > <img alt="image" width="1204" src="https://private-user-images.githubusercontent.com/38887641/335924670-5b28858d-6b91-4d87-9df8-6e60a4d9c621.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczODQ5MTcsIm5iZiI6MTcxNzM4NDYxNywicGF0aCI6Ii8zODg4NzY0MS8zMzU5MjQ2NzAtNWIyODg1OGQtNmI5MS00ZDg3LTlkZjgtNmU2MGE0ZDljNjIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjAzVDAzMTY1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5MzBkYzZhNTM1ZjhiODc2ZWVhYmFiZWI5ZTBkMzBmNjA3YWQzMGM3NTc0MmVmYzk0MjQ5ODMyMmJjNzM2YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sod7wWS9iKKcqY_RHAm8PaimUDxCd9noI-Zxux98y3c"> golang 里面时间统计，用 time.Since 好点，使用 时间戳相减，有时候会出现负数，之前我们生产环境遇过这个坑。不知道新版本的 golang 是否解决了这个问题。即，r.ReceiveTime 用 time.Time 来表达，不要使用时间戳
> > > codis没有遇到过
> 
> 按照月财这个方式处理


Golang 1.9Monotonic Clock在软件包中引入了该功能time，解决了由于时钟调整（例如，NTP 更新）而减去时间戳可能导致负持续时间的问题